File: mooncake-ep/benchmarks/legacy_buffer_perf.cpp
#include <cuda_alike.h>

#include <algorithm>
#include <cstdint>
#include <stdexcept>
#include <string>
#include <vector>

#include <mooncake_ep_buffer.h>
#include <pybind11/gil.h>
#include <pybind11/pybind11.h>

namespace py = pybind11;

namespace mooncake {
namespace {

struct LegacyBufferPerfTensors {
    uint64_t x_ptr = 0;
    uint64_t topk_idx_ptr = 0;
    uint64_t topk_weights_ptr = 0;
    uint64_t active_ranks_ptr = 0;
    uint64_t expert_x_ptr = 0;
    uint64_t packed_recv_x_ptr = 0;
    uint64_t packed_recv_count_ptr = 0;
    uint64_t packed_recv_src_info_ptr = 0;
    uint64_t packed_recv_layout_range_ptr = 0;
    uint64_t combined_x_ptr = 0;
};

struct LegacyBufferPerfConfig {
    int num_tokens = 0;
    int hidden = 0;
    int num_topk = 0;
    int num_local_experts = 0;
    int num_max_dispatch_tokens_per_rank = 0;
    int num_experts = 0;
    int timeout_us = -1;
    int warmups = 20;
    int iterations = 30;
    uint64_t compute_stream_ptr = 0;
};

struct LegacyBufferPerfResult {
    double average_us = 0.0;
    double min_us = 0.0;
    double max_us = 0.0;
    double dispatch_average_us = 0.0;
    double combine_average_us = 0.0;
};

void cuda_check(cudaError_t status, const char* operation) {
    if (status != cudaSuccess) {
        throw std::runtime_error(std::string(operation) + ": " +
                                 cudaGetErrorString(status));
    }
}

void validate(MooncakeEpBuffer& buffer, const LegacyBufferPerfTensors& tensors,
              const LegacyBufferPerfConfig& config) {
    if (config.num_tokens <= 0 || config.hidden <= 0 || config.num_topk <= 0 ||
        config.num_max_dispatch_tokens_per_rank < config.num_tokens ||
        config.num_experts <= 0 || config.num_local_experts <= 0 ||
        config.warmups < 0 || config.iterations <= 1) {
        throw std::invalid_argument(
            "invalid legacy EP benchmark configuration");
    }
    if (buffer.ibgda_disabled() && !buffer.use_fast_path()) {
        throw std::runtime_error(
            "legacy EP benchmark requires the native fast path");
    }
    if (!tensors.x_ptr || !tensors.topk_idx_ptr || !tensors.topk_weights_ptr ||
        !tensors.active_ranks_ptr || !tensors.expert_x_ptr ||
        !tensors.packed_recv_x_ptr || !tensors.packed_recv_count_ptr ||
        !tensors.packed_recv_src_info_ptr ||
        !tensors.packed_recv_layout_range_ptr || !tensors.combined_x_ptr) {
        throw std::invalid_argument(
            "legacy EP benchmark received a null tensor pointer");
    }
}

LegacyBufferPerfResult run_legacy_buffer_perf(
    MooncakeEpBuffer& buffer, const LegacyBufferPerfTensors& tensors,
    const LegacyBufferPerfConfig& config) {
    validate(buffer, tensors, config);

    const int num_local_experts = config.num_local_experts;
    const auto compute_stream =
        reinterpret_cast<cudaStream_t>(config.compute_stream_ptr);
    const size_t recv_count_bytes =
        static_cast<size_t>(num_local_experts) * sizeof(int);
#ifdef MOONCAKE_EP_USE_MUSA
    // MUSA split SEND/RECV uses a communication stream. Avoid the synchronous
    // compute<->communication stream cycle in the generic legacy wrapper; the
    // benchmark synchronizes the device before recording each end event.
    constexpr bool kAsyncDispatch = true;
    constexpr bool kAsyncCombine = false;
#else
    constexpr bool kAsyncDispatch = false;
    constexpr bool kAsyncCombine = false;
#endif

    auto run_dispatch = [&] {
        cuda_check(cudaMemsetAsync(
                       reinterpret_cast<void*>(tensors.packed_recv_count_ptr),
                       0, recv_count_bytes, compute_stream),
                   "cudaMemsetAsync(packed_recv_count)");
        buffer.dispatch(
            tensors.x_ptr, tensors.topk_idx_ptr, tensors.active_ranks_ptr,
            config.num_tokens, config.hidden, config.num_topk,
            config.num_max_dispatch_tokens_per_rank, config.num_experts,
            config.timeout_us, false, tensors.packed_recv_x_ptr, 0,
            tensors.packed_recv_count_ptr, tensors.packed_recv_src_info_ptr,
            tensors.packed_recv_layout_range_ptr, kAsyncDispatch, false,
            config.compute_stream_ptr);
    };
    auto run_combine = [&] {
        buffer.combine(
            tensors.expert_x_ptr, tensors.topk_idx_ptr,
            tensors.topk_weights_ptr, tensors.packed_recv_src_info_ptr,
            tensors.packed_recv_layout_range_ptr, tensors.active_ranks_ptr,
            num_local_experts, config.num_tokens, config.hidden,
            config.num_topk, config.num_max_dispatch_tokens_per_rank,
            config.num_experts, config.timeout_us, false,
            tensors.combined_x_ptr, kAsyncCombine, false,
            config.compute_stream_ptr);
    };
    auto run_once = [&] {
        run_dispatch();
        run_combine();
    };

    for (int i = 0; i < config.warmups; ++i) run_once();
    cuda_check(cudaDeviceSynchronize(), "cudaDeviceSynchronize(warmup)");

    std::vector<cudaEvent_t> starts(config.iterations);
    std::vector<cudaEvent_t> ends(config.iterations);
    std::vector<cudaEvent_t> dispatch_starts(config.iterations);
    std::vector<cudaEvent_t> dispatch_ends(config.iterations);
    std::vector<cudaEvent_t> combine_starts(config.iterations);
    std::vector<cudaEvent_t> combine_ends(config.iterations);
    for (int i = 0; i < config.iterations; ++i) {
        cuda_check(cudaEventCreate(&starts[i]), "cudaEventCreate(start)");
        cuda_check(cudaEventCreate(&ends[i]), "cudaEventCreate(end)");
        cuda_check(cudaEventCreate(&dispatch_starts[i]), "cudaEventCreate(dispatch_start)");
        cuda_check(cudaEventCreate(&dispatch_ends[i]), "cudaEventCreate(dispatch_end)");
        cuda_check(cudaEventCreate(&combine_starts[i]), "cudaEventCreate(combine_start)");
        cuda_check(cudaEventCreate(&combine_ends[i]), "cudaEventCreate(combine_end)");
        cuda_check(cudaEventRecord(starts[i], compute_stream),
                   "cudaEventRecord(start)");
        cuda_check(cudaEventRecord(dispatch_starts[i], compute_stream),
                   "cudaEventRecord(dispatch_start)");
        run_dispatch();
        cuda_check(cudaDeviceSynchronize(), "cudaDeviceSynchronize(dispatch)");
        cuda_check(cudaEventRecord(dispatch_ends[i], compute_stream),
                   "cudaEventRecord(dispatch_end)");
        cuda_check(cudaEventRecord(combine_starts[i], compute_stream),
                   "cudaEventRecord(combine_start)");
        run_combine();
        cuda_check(cudaDeviceSynchronize(), "cudaDeviceSynchronize(iteration)");
        cuda_check(cudaEventRecord(combine_ends[i], compute_stream),
                   "cudaEventRecord(combine_end)");
        cuda_check(cudaEventRecord(ends[i], compute_stream),
                   "cudaEventRecord(end)");
    }
    cuda_check(cudaEventSynchronize(ends.back()), "cudaEventSynchronize(end)");

    std::vector<float> timings_us;
    timings_us.reserve(config.iterations - 1);
    for (int i = 1; i < config.iterations; ++i) {
        float elapsed_ms = 0.0f;
        cuda_check(cudaEventElapsedTime(&elapsed_ms, starts[i], ends[i]),
                   "cudaEventElapsedTime");
        timings_us.push_back(elapsed_ms * 1000.0f);
    }
    for (auto event : starts) cudaEventDestroy(event);
    for (auto event : ends) cudaEventDestroy(event);

    std::vector<float> dispatch_timings_us;
    std::vector<float> combine_timings_us;
    dispatch_timings_us.reserve(config.iterations - 1);
    combine_timings_us.reserve(config.iterations - 1);
    for (int i = 1; i < config.iterations; ++i) {
        float elapsed_ms = 0.0f;
        cuda_check(cudaEventElapsedTime(&elapsed_ms, dispatch_starts[i], dispatch_ends[i]),
                   "cudaEventElapsedTime(dispatch)");
        dispatch_timings_us.push_back(elapsed_ms * 1000.0f);
        elapsed_ms = 0.0f;
        cuda_check(cudaEventElapsedTime(&elapsed_ms, combine_starts[i], combine_ends[i]),
                   "cudaEventElapsedTime(combine)");
        combine_timings_us.push_back(elapsed_ms * 1000.0f);
    }
    for (auto event : dispatch_starts) cudaEventDestroy(event);
    for (auto event : dispatch_ends) cudaEventDestroy(event);
    for (auto event : combine_starts) cudaEventDestroy(event);
    for (auto event : combine_ends) cudaEventDestroy(event);

    double total_us = 0.0;
    for (float value : timings_us) total_us += value;
    const auto [min_it, max_it] =
        std::minmax_element(timings_us.begin(), timings_us.end());
    double dispatch_total_us = 0.0;
    double combine_total_us = 0.0;
    for (size_t i = 0; i < dispatch_timings_us.size(); ++i) {
        dispatch_total_us += dispatch_timings_us[i];
        combine_total_us += combine_timings_us[i];
    }
    return {total_us / timings_us.size(), *min_it, *max_it,
            dispatch_total_us / dispatch_timings_us.size(),
            combine_total_us / combine_timings_us.size()};
}

}  // namespace

void bind_legacy_buffer_perf(py::module_& module) {
    py::class_<LegacyBufferPerfTensors>(module, "_LegacyBufferPerfTensors")
        .def(py::init<>())
        .def_readwrite("x_ptr", &LegacyBufferPerfTensors::x_ptr)
        .def_readwrite("topk_idx_ptr", &LegacyBufferPerfTensors::topk_idx_ptr)
        .def_readwrite("topk_weights_ptr",
                       &LegacyBufferPerfTensors::topk_weights_ptr)
        .def_readwrite("active_ranks_ptr",
                       &LegacyBufferPerfTensors::active_ranks_ptr)
        .def_readwrite("expert_x_ptr", &LegacyBufferPerfTensors::expert_x_ptr)
        .def_readwrite("packed_recv_x_ptr",
                       &LegacyBufferPerfTensors::packed_recv_x_ptr)
        .def_readwrite("packed_recv_count_ptr",
                       &LegacyBufferPerfTensors::packed_recv_count_ptr)
        .def_readwrite("packed_recv_src_info_ptr",
                       &LegacyBufferPerfTensors::packed_recv_src_info_ptr)
        .def_readwrite("packed_recv_layout_range_ptr",
                       &LegacyBufferPerfTensors::packed_recv_layout_range_ptr)
        .def_readwrite("combined_x_ptr",
                       &LegacyBufferPerfTensors::combined_x_ptr);

    py::class_<LegacyBufferPerfConfig>(module, "_LegacyBufferPerfConfig")
        .def(py::init<>())
        .def_readwrite("num_tokens", &LegacyBufferPerfConfig::num_tokens)
        .def_readwrite("hidden", &LegacyBufferPerfConfig::hidden)
        .def_readwrite("num_topk", &LegacyBufferPerfConfig::num_topk)
        .def_readwrite("num_local_experts",
                       &LegacyBufferPerfConfig::num_local_experts)
        .def_readwrite(
            "num_max_dispatch_tokens_per_rank",
            &LegacyBufferPerfConfig::num_max_dispatch_tokens_per_rank)
        .def_readwrite("num_experts", &LegacyBufferPerfConfig::num_experts)
        .def_readwrite("timeout_us", &LegacyBufferPerfConfig::timeout_us)
        .def_readwrite("warmups", &LegacyBufferPerfConfig::warmups)
        .def_readwrite("iterations", &LegacyBufferPerfConfig::iterations)
        .def_readwrite("compute_stream_ptr",
                       &LegacyBufferPerfConfig::compute_stream_ptr);

    py::class_<LegacyBufferPerfResult>(module, "_LegacyBufferPerfResult")
        .def_readonly("average_us", &LegacyBufferPerfResult::average_us)
        .def_readonly("min_us", &LegacyBufferPerfResult::min_us)
        .def_readonly("max_us", &LegacyBufferPerfResult::max_us)
        .def_readonly("dispatch_average_us",
                      &LegacyBufferPerfResult::dispatch_average_us)
        .def_readonly("combine_average_us",
                      &LegacyBufferPerfResult::combine_average_us);

    module.def("_benchmark_legacy_buffer", &run_legacy_buffer_perf,
               py::arg("buffer"), py::arg("tensors"), py::arg("config"),
               py::call_guard<py::gil_scoped_release>());
}

}  // namespace mooncake

File: mooncake-ep/benchmarks/legacy_buffer_perf.py
#!/usr/bin/env python3
"""Native-core legacy EP benchmark with Python distributed bootstrap.

Launch with torchrun. Python initializes the Mooncake process group and owns
the fixed tensor storage; the timed dispatch/combine loop executes entirely in
the torch-free C++ EP module.
"""

import argparse
import os

import torch
import torch.distributed as dist

import mooncake._ep as native_ep
import mooncake.pg  # Registers the Mooncake process-group backend for bootstrap.
from mooncake.mooncake_ep_buffer import Buffer


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--tokens", type=int, default=128)
    parser.add_argument("--hidden", type=int, default=7168)
    parser.add_argument("--experts", type=int, default=288)
    parser.add_argument("--topk", type=int, default=8)
    parser.add_argument("--warmups", type=int, default=20)
    parser.add_argument("--iterations", type=int, default=30)
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    local_rank = int(os.environ["LOCAL_RANK"])
    torch.cuda.set_device(local_rank)
    dist.init_process_group("mooncake")
    world_size = dist.get_world_size()
    rank = dist.get_rank()
    group = dist.new_group(list(range(world_size)))

    if args.experts % world_size:
        raise ValueError("--experts must be divisible by world size")
    if args.hidden % 128:
        raise ValueError("--hidden must be divisible by 128")

    torch.manual_seed(rank)
    num_local_experts = args.experts // world_size
    buffer_bytes = Buffer.get_ep_buffer_size_hint(
        args.tokens, args.hidden, world_size, args.experts
    )
    buffer = Buffer(group, num_ep_buffer_bytes=buffer_bytes)
    if buffer._use_fallback:
        raise RuntimeError("native-core benchmark requires the EP fast path")

    x = torch.randn(
        (args.tokens, args.hidden), dtype=torch.bfloat16, device="cuda"
    )
    scores = torch.randn(
        (args.tokens, args.experts), dtype=torch.float32, device="cuda"
    )
    topk_idx = torch.topk(scores, args.topk, dim=-1).indices.contiguous()
    topk_weights = torch.rand(
        (args.tokens, args.topk), dtype=torch.float32, device="cuda"
    )
    active_ranks = torch.ones(world_size, dtype=torch.int32, device="cuda")
    recv_tokens = world_size * args.tokens
    expert_x = torch.randn(
        (num_local_experts, recv_tokens, args.hidden),
        dtype=torch.bfloat16,
        device="cuda",
    )
    packed_recv_x = torch.empty_like(expert_x)
    packed_recv_count = torch.empty(
        num_local_experts, dtype=torch.int32, device="cuda"
    )
    packed_recv_src_info = torch.empty(
        (num_local_experts, recv_tokens), dtype=torch.int32, device="cuda"
    )
    packed_recv_layout_range = torch.empty(
        (num_local_experts, world_size), dtype=torch.int64, device="cuda"
    )
    combined_x = torch.empty_like(x)
    cache_flush = torch.empty(int(256e6 // 4), dtype=torch.int32, device="cuda")

    tensors = native_ep._LegacyBufferPerfTensors()
    tensors.x_ptr = x.data_ptr()
    tensors.topk_idx_ptr = topk_idx.data_ptr()
    tensors.topk_weights_ptr = topk_weights.data_ptr()
    tensors.active_ranks_ptr = active_ranks.data_ptr()
    tensors.expert_x_ptr = expert_x.data_ptr()
    tensors.packed_recv_x_ptr = packed_recv_x.data_ptr()
    tensors.packed_recv_count_ptr = packed_recv_count.data_ptr()
    tensors.packed_recv_src_info_ptr = packed_recv_src_info.data_ptr()
    tensors.packed_recv_layout_range_ptr = packed_recv_layout_range.data_ptr()
    tensors.combined_x_ptr = combined_x.data_ptr()

    config = native_ep._LegacyBufferPerfConfig()
    config.num_tokens = args.tokens
    config.hidden = args.hidden
    config.num_topk = args.topk
    config.num_local_experts = num_local_experts
    config.num_max_dispatch_tokens_per_rank = args.tokens
    config.num_experts = args.experts
    config.warmups = args.warmups
    config.iterations = args.iterations
    config.compute_stream_ptr = torch.cuda.current_stream().cuda_stream

    dist.barrier(group=group)
    cache_flush.zero_()
    result = native_ep._benchmark_legacy_buffer(buffer.runtime, tensors, config)
    selections = topk_idx.numel()
    payload_bytes = selections * (args.hidden * 4)
    bandwidth = payload_bytes / result.average_us / 1e3
    print(
        f"[rank {rank}] Native dispatch + combine: {bandwidth:.2f} GB/s, "
        f"avg_t={result.average_us:.2f} us, min_t={result.min_us:.2f} us, "
        f"max_t={result.max_us:.2f} us",
        flush=True,
    )
    dist.destroy_process_group()


if __name__ == "__main__":
    main()
